Common/Securities/BuyingPowerModel.cs

```diff
@@ -358,141 +358,113 @@ public virtual HasSufficientBuyingPowerForOrderResult HasSufficientBuyingPowerFo
-            var currentSignedUsedMargin = this.GetInitialMarginRequirement(parameters.Security, parameters.Security.Holdings.Quantity);
-
-            // remove directionality, we'll work in the land of absolutes
-            var absFinalOrderMargin = Math.Abs(signedTargetFinalMarginValue - currentSignedUsedMargin);
-            var direction = signedTargetFinalMarginValue > currentSignedUsedMargin ? OrderDirection.Buy : OrderDirection.Sell;
+            // Determine how much of our portfolio is useable for this target; TPV * (1 - BufferPercent)
+            var totalUseablePortfolioValue = parameters.Portfolio.TotalPortfolioValue * (1 - RequiredFreeBuyingPowerPercent);
 
-            // determine the unit price in terms of the account currency
-            var utcTime = parameters.Security.LocalTime.ConvertToUtc(parameters.Security.Exchange.TimeZone);
-            // determine the margin required for 1 unit, positive since we are working with absolutes
-            var absUnitMargin = this.GetInitialMarginRequirement(parameters.Security, 1);
-            if (absUnitMargin == 0)
+            // Our max target margin allocated to this security; if targeting zero, simply return the negative of the quantity
+            var targetMarginAllocated = parameters.TargetBuyingPower * totalUseablePortfolioValue;
+            if (targetMarginAllocated == 0)
             {
-                return new GetMaximumOrderQuantityResult(0, parameters.Security.Symbol.GetZeroPriceMessage());
+                return new GetMaximumOrderQuantityResult(-parameters.Security.Holdings.Quantity, string.Empty,
+                    false);
             }
 
-            var minimumValue = absUnitMargin * parameters.Security.SymbolProperties.LotSize;
-            if (minimumValue > absFinalOrderMargin)
-            {
-                string reason = null;
-                if (!parameters.SilenceNonErrorReasons)
-                {
-                    reason = $"The target order margin {absFinalOrderMargin} is less than the minimum {minimumValue}.";
-                }
-                return new GetMaximumOrderQuantityResult(0, reason, false);
-            }
-
-            // continue iterating while we do not have enough margin for the order
-            decimal orderMargin = 0;
-            decimal orderFees = 0;
-            // compute the initial order quantity
-            var orderQuantity = absFinalOrderMargin / absUnitMargin;
+            // First pass, calculate:
+            // - Final holdings quantity
+            // - Order quantity to get to our final holdings
+            // - Total target holdings margin requirement
+            var finalHoldingsQuantity = (targetMarginAllocated / unitMargin)
+                .DiscretelyRoundBy(parameters.Security.SymbolProperties.LotSize);
+            var orderQuantity = finalHoldingsQuantity - parameters.Security.Holdings.Quantity;
+            var targetHoldingsMargin = finalHoldingsQuantity * unitMargin;
```

Comment:
`finalHoldingsQuantity * unitMargin` what about lotSize for the resulting operation?

Reply:
Not needed for that guy, he is a dollar value for the target holdings margin used if we were to have finalHoldingsQuantity. 😁